app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/ui/onboarding/VpnStore.kt

```diff
@@ -93,6 +102,26 @@ class SharedPreferencesVpnStore @Inject constructor(
         return preferences.getBoolean(KEY_ALWAYS_ON_MODE_ENABLED, false)
     }
 
+    override suspend fun vpnLastDisabledByAndroid(): Boolean {
+        fun vpnUnexpectedlyDisabled(): Boolean {
+            return vpnServiceStateDao.getLastStateStats()?.let {
+                (
+                    it.state == VpnServiceState.DISABLED &&
+                        it.stopReason != VpnStoppingReason.SELF_STOP &&
+                        it.stopReason != VpnStoppingReason.REVOKED
+                    )
+            } ?: false
+        }
+
+        fun vpnKilledBySystem(): Boolean {
+            val lastHeartBeat = vpnHeartBeatDao.hearBeats().maxByOrNull { it.timestamp }
+            return lastHeartBeat?.type == VpnServiceHeartbeatMonitor.DATA_HEART_BEAT_TYPE_ALIVE &&
+                !vpnFeaturesRegistry.isFeatureRegistered(AppTpVpnFeature.APPTP_VPN)
+        }
```

Comment:
Internal `fun` because I don't want to call them outside the main fun

Reply:
that would be no fun